scylla/src/transport/locator/tablets.rs

```diff
@@ -0,0 +1,901 @@
+            first_token: Token::new(first_token + 1),
+            last_token: Token::new(last_token),
+            replicas: RawTabletReplicas { replicas },
+        }))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+struct TabletReplicas {
+    all: Vec<(Arc<Node>, Shard)>,
+    per_dc: HashMap<String, Vec<(Arc<Node>, Shard)>>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Tablet {
+    /// First token belonging to the tablet, inclusive
+    first_token: Token,
+    /// Last token belonging to the tablet, inclusive
+    last_token: Token,
+    replicas: TabletReplicas,
+    /// If any of the replicas failed to resolve to a Node,
+    /// then this field will contain the original list of replicas.
+    failed: Option<RawTabletReplicas>,
+}
+
+/// Container for tablets of a single table.
+///
+/// It can be viewed as a set of non-overlapping Tablet objects.
+/// It has 2 basic operations:
+/// 1. Find a tablet for given Token
+/// 2. Add a new tablet.
+///
+/// Adding new Tablet will first remove all tablets that overlap with the new tablet.
+#[derive(Clone, Debug)]
+pub(crate) struct TableTablets {
+    table_spec: TableSpec<'static>,
+    tablet_list: Vec<Tablet>,
+    /// In order to make typical tablet maintance faster
+    /// we remember if there were any tablets that have unrecognized uuids in replica list.
+    /// If there were none, and a few other conditions are satisfied, we can skip nearly whole maintanace.
+    /// This flag may be falsely true: if we add tablet with unknown replica but later
+    /// overwrite it with some other tablet.
+    has_unknown_replicas: bool,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct TabletsInfo {
+    tablets: hashbrown::HashMap<TableSpec<'static>, TableTablets>,
+    /// See `has_unknown_replicas` field in `TableTablets`.
+    /// The field here will be true if it is true for any TableTablets.
+    has_unknown_replicas: bool,
+}
+
+impl Tablet {
+    pub(crate) fn from_raw_tablet(
+        raw_tablet: RawTablet,
+        replica_translator: impl Fn(Uuid) -> Option<Arc<Node>>,
+    ) -> Result<Self, (Self, Vec<Uuid>)> {
+        let replicas_result =
+            TabletReplicas::from_raw_replicas(&raw_tablet.replicas, replica_translator);
+        match replicas_result {
+            Ok(replicas) => Ok(Self {
+                first_token: raw_tablet.first_token,
+                last_token: raw_tablet.last_token,
+                replicas,
+                failed: None,
+            }),
+            Err((replicas, failed_replicas)) => Err((
+                Self {
+                    first_token: raw_tablet.first_token,
+                    last_token: raw_tablet.last_token,
+                    replicas,
+                    failed: Some(raw_tablet.replicas),
+                },
+                failed_replicas,
+            )),
+        }
+    }
+
+    pub(crate) fn range(&self) -> (Token, Token) {
+        (self.first_token, self.last_token)
+    }
```

Comment:
This is not a range meant for iteration, just for printing a debug info. I don't see how this change would improve the code, but it makes it more complicated.

Reply:
`Range` object is not only intended for iteration, but it also provides `start()` and `end()` methods, too. For me it would make a great sense to return `Range` object from `range` method.
But OK, it's a nit.